### PR TITLE
Api attach detach partition

### PR DESCRIPTION
**Pull request by pyakushin**

### CODE REVIEW

File: src/api/http.rs [hunk 1]
```diff
@@ -219,17 +210,29 @@ fn partition_by_id(
         })
 }
 
-#[put("/vdisks/<vdisk_id>/partitions/<partition_id>/<action>")]
+#[get("/vdisks/<vdisk_id>/partitions/<partition_id>/<action>")]
```
Comment (agend) [inline src/api/http.rs]: any change should not be GET it should be `POST`  or `PUT`  or `PATCH`. Here lets use `PUT` as we replacing state

Comment (agend) [reply src/api/http.rs]: Let make it `POST`)

File: src/api/http.rs [hunk 2]
```diff
@@ -219,17 +210,29 @@ fn partition_by_id(
...
+        "partition with id: {} in vdisk {} is successfully {:?}ed",
+        partition_id, vdisk_id, action
+    );
+    let task = async move {
+        match action {
+            Action::Attach => group.attach(partition_id).await,
+            Action::Detach => group.detach(partition_id).await,
+        }
+    };
+    rt.block_on(task);
```
Comment (agend) [inline src/api/http.rs]: why `block_on` here? can we go with `await`?

File: src/api/http.rs [hunk 3]
```diff
@@ -219,17 +210,29 @@ fn partition_by_id(
...
-    format!(
-        "{:?} partittion {} of vd {}",
-        action, partition_id, vdisk_id
-    )
+) -> Result<StatusExt, StatusExt> {
+    let group = find_group(&bob, vdisk_id)?;
+    let group = group.clone();
+    let mut rt = Runtime::new().expect("create runtime");
+    let res = format!(
+        "partition with id: {} in vdisk {} is successfully {:?}ed",
```
Comment (agend) [inline src/api/http.rs]: better to come up with json result message. example: { ok: true, msg: "text" }

File: src/api/http.rs (hunk 1 above)
Comment (pyakushin) [reply src/api/http.rs]: Fixed

File: src/api/http.rs (hunk 3 above)
Comment (pyakushin) [reply src/api/http.rs]: Done

File: src/api/http.rs (hunk 2 above)
Comment (pyakushin) [reply src/api/http.rs]: Api server runs in its own thread. And runtime is required for only `change_partition_state`, for now. So in this case it is easier to block on a task than to solve problems with launching rocket server with our own tokio runtime.

**Comment by agend** [comment]
After detaching got this `ERROR`  in log: 
`2019-12-05T23:52:32.307199634+03:00 ERROR pearl::storage::observer - Inner(ActiveBlobNotSet)`

but all successful. Is this ok?

**Comment by agend** [comment]
Getting a lot of this while benchmarking:
```
2019-12-05T23:53:08.249654885+03:00 WARN dipstick::output::graphite - Could not flush graphite metrics upon Drop: not connected
2019-12-05T23:53:08.250122526+03:00 WARN dipstick::output::graphite - Could not flush graphite metrics upon Drop: not connected
2019-12-05T23:53:08.250660245+03:00 WARN dipstick::output::graphite - Could not flush graphite metrics upon Drop: not connected
2019-12-05T23:53:08.251052695+03:00 WARN dipstick::output::graphite - Could not flush graphite metrics upon Drop: not connected
2019-12-05T23:53:08.251360755+03:00 WARN dipstick::output::graphite - Could not flush graphite metrics upon Drop: not connected
2019-12-05T23:53:08.251634370+03:00 WARN dipstick::output::graphite - Could not flush graphite metrics upon Drop: not connected
2019-12-05T23:53:08.251917546+03:00 WARN dipstick::output::graphite - Could not flush graphite metrics upon Drop: not connected
2019-12-05T23:53:08.252189742+03:00 WARN dipstick::output::graphite - Could not flush graphite metrics upon Drop: not connected
2019-12-05T23:53:08.252478912+03:00 WARN dipstick::output::graphite - Could not flush graphite metrics upon Drop: not connected
2019-12-05T23:53:08.252748942+03:00 WARN dipstick::output::graphite - Could not flush graphite metrics upon Drop: not connected
2019-12-05T23:53:08.253019252+03:00 WARN dipstick::output::graphite - Could not flush graphite metrics upon Drop: not connected
2019-12-05T23:53:08.253303209+03:00 WARN dipstick::output::graphite - Could not flush graphite metrics upon Drop: not connected
2019-12-05T23:53:08.253514737+03:00 WARN dipstick::output::graphite - Could not flush graphite metrics upon Drop: not connected
2019-12-05T23:53:08.253654846+03:00 ERROR dipstick::core::scheduler - Could not flush metrics: not connected
```

**Comment by agend** [comment]
> After detaching got this `ERROR` in log:
> `2019-12-05T23:52:32.307199634+03:00 ERROR pearl::storage::observer - Inner(ActiveBlobNotSet)`
> 
> but all successful. Is this ok?

@pyakushin Probably we should not allow detaching partition with active blob in it

**Comment by agend** [comment]
> > After detaching got this `ERROR` in log:
> > `2019-12-05T23:52:32.307199634+03:00 ERROR pearl::storage::observer - Inner(ActiveBlobNotSet)`
> > but all successful. Is this ok?
> 
> @pyakushin Probably we should not allow detaching partition with active blob in it

After detach i wrote more data - no error it just reused folder.

**Comment by pyakushin** [comment]
> After detaching got this `ERROR` in log:
> `2019-12-05T23:52:32.307199634+03:00 ERROR pearl::storage::observer - Inner(ActiveBlobNotSet)`
> 
> but all successful. Is this ok?

It happens when storage closes active blob before observer stops

File: src/core_inner/backend/pearl/group.rs [hunk 4]
```diff
@@ -301,25 +301,44 @@ impl PearlGroup {
...
+                let storage: &Storage<_> = pearl_sync.storage.as_ref().expect("pearl storage");
+                if let Err(e) = storage.close().await {
+                    warn!("pearl closed: {:?}", e);
+                }
+                pearls.retain(|pearl| pearl.start_timestamp != start_timestamp);
+                Ok(())
             }
+        } else {
+            let msg = format!("pearl:{} not found", start_timestamp);
+            Err(Error::PearlChangeState(msg))
```
Comment (agend) [inline src/core_inner/backend/pearl/group.rs]: now message from server looks like: { ok: false, msg: PearlChangeState("current active pearl:1576022400 cannot be detached")  } it's not valid JSON.

Comment (pyakushin) [reply src/core_inner/backend/pearl/group.rs]: Fixed: `{ ok: false, msg: "backend error: current active pearl:1576627200 cannot be detached" }`

**Comment by agend** [comment]
Ok and msg should be in quotes as well: check with json validator

**Pull request merged**